---
title: "Updated Notes"
author: "Jialin Ma"
date: "June 14, 2017"
output:
    rmarkdown::html_document:
        toc: true
        toc_float: true
        theme: flatly
---


In this document, I aim to showcase the current available features of TnT package and
briefly describe the structure and compilation process of TnT Board and TnT Tracks.


# Usage

## Install

To install the version of package that builds this document, use:

```{r, eval=FALSE}
install_github("marlin-na/TnT", ref = "93cd102e7b970e4d4df0850d2edcac26867d4355")
```

```{r}
suppressPackageStartupMessages({
    library(TnT)
    library(GenomicFeatures)
})
```


## Sample Data

### GRanges Data

cpgIslands is a dummy GRanges object from Gviz package.

```{r}
data(cpgIslands, package = "Gviz")
cpgIslands
```

Here I also supply a custom defined GRanges object with multiple chromosomes.

```{r}
ir <- IRanges(start = seq(26000000, 30000000, by = 100000), width = 50000)
myRange <- c(
    GRanges("chr7", ir),
    GRanges("chr8", shift(ir, 5000))
)
myRange <- myRange[order(start(myRange))]
myRange
```

### Width-one GRanges with Paired Values

Dummy position-value pairs.

```{r}
myPos <- GRanges("chr7", IRanges(start = sample(26000000:27000000, 100), width = 1))
myPos$value <- runif(n = length(myPos))
myPos
```


### Genome Annotation

The transcript database for human genome.

```{r}
hg19.txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene::TxDb.Hsapiens.UCSC.hg19.knownGene
hg19.txdb
```


## The Browser Instance

Tracks are initiated by cooresponding functions with data.

```{r}
# Tracks
gene_track  <- GeneTrack(hg19.txdb, color = "grey", height = 200)
pin_track   <- PinTrack(myPos, color = "red", height = 50)
block_track <- BlockTrack(myRange, color = "brown", height = 50)
cpg_track   <- BlockTrack(cpgIslands, color = "blue", height = 50)
```

The TnT Board is constructed with a list of tracks, and a view range to indicate the chromosome and range to view.

```{r}
board <- TnTBoard(
    viewrange = GRanges("chr7", IRanges(26000000, 27000000)),
    
    list(gene_track, pin_track,
         cpg_track, block_track)
)
# TODO: should be simplied
TnT(compileBoard(board))
```


# Explanation

## Structure of TnT Track

A TnT Track consists of three slots:

```{r}
showClass("TnTTrack")
```

- Spec:
    Class "TrackSpec" with three slots (i.e. "background", "height", "label").
    It describes the common options across tracks (background color, height
    and label for each track).

- Data:
    Class "TrackData", it is a virtual class with subclasses that have different
    methods for conversion to JavaScript. Currently all subclasses also extend
    "GRanges" class.

- Display:
    A list, describes the display options for different types of track,
    e.g. color of the feature, height "domain" of AreaTrack, and in future, will
    also include the settings for tooltips

TnT Track supports `seqinfo` and `seqinfo<-` method so that before compilation,
only track data on the intended chromosome can be kept. For example:

```{r}
seqlevels(block_track)
seqlevels(
    dropSeqlevels(block_track, "chr8", pruning.mode = "coarse")
)
```

## Structure of TnT Board

```{r}
showClass("TnTBoard")
```

A TnT Board is the object that holds a list of tracks and settings for the browser,
it contains the following slots:

- ViewRange:
    Class "GRanges" that specifies the initial viewing range of rendered browser.
    According to the its seqlevel, trackdata that on extra chromosomes will be
    dropped before converted to JavaScript. It should be suppiled when constructing
    the Board.
    
- CoordRange:
    Class "IRanges" that defines the left and right coordinate of the board.
    It is not suppied when constructing the board.
    But before compilation, it is determined by the "seqlength" of "seqinfo"
    from ViewRange and also the track list, if not found, use value of
    minimal and maximal range of track data list.
    
- ZoomAllow:
    Class "IRanges" that describes the minimal and maximal extent of the
    board (i.e. the limit when zooming in and out).
    Like CoordRange, it is not set when constructing board.
    Before compilation, the maximal extent is set to width of CoordRange, the
    minimal extent is set to 10.
    
- TrackList
    List of tracks.

## Compilation Process

In summary, before compilation, TnT should do the following work:

1. Drop trackdata that are not on the seqlevel (chromosome) of ViewRange.
2. Determine the coordinate range and maximal limit of zooming
3. Check TrackSpec of each track, fill with default values if they are not specified (background color and height of the track).

Finally, the TnTBoard object can be compiled to a valid definition of TnT instance, which is relative
straightforward.

# TODO

Further, I should refine this framework and extend it with more track types including
the transcript track, line track, area track, and switch to use tnt.genome.

## Tasks

1. Provide transcript track and provide other types of track.
2. Display tooltips.
3. The size/width of the board should to be responsive, by using the resize method for htmlwidget.
4. Implement TnT Genome class, which should inherit the TnT Board class.
5. Improve the printing method of TnTBoard, so that it can be directly rendered under shiny and rmarkdown.
6. Informative show method for tracks and track data.






